pkg/library/mappers/httproute.go

```diff
@@ -20,25 +25,63 @@ type httpRouteEventMapper struct {
 	opts MapperOptions
 }
 
-func (m *httpRouteEventMapper) MapToPolicy(obj client.Object, policyKind kuadrant.Referrer) []reconcile.Request {
+func (m *httpRouteEventMapper) MapToPolicy(ctx context.Context, obj client.Object, policyKind kuadrantgatewayapi.Policy) []reconcile.Request {
 	logger := m.opts.Logger.WithValues("httproute", client.ObjectKeyFromObject(obj))
-
+	requests := make([]reconcile.Request, 0)
 	httpRoute, ok := obj.(*gatewayapiv1.HTTPRoute)
 	if !ok {
 		logger.Info("cannot map httproute event to kuadrant policy", "error", fmt.Sprintf("%T is not a *gatewayapiv1beta1.HTTPRoute", obj))
 		return []reconcile.Request{}
 	}
 
-	requests := make([]reconcile.Request, 0)
+	gatewayKeys := kuadrantgatewayapi.GetRouteAcceptedGatewayParentKeys(httpRoute)
 
-	for _, policyKey := range kuadrant.BackReferencesFromObject(httpRoute, policyKind) {
-		logger.V(1).Info("kuadrant policy possibly affected by the httproute related event found", policyKind.Kind(), policyKey)
-		requests = append(requests, reconcile.Request{NamespacedName: policyKey})
+	for _, gatewayKey := range gatewayKeys {
+		gateway := &gatewayapiv1.Gateway{}
+		err := m.opts.Client.Get(ctx, gatewayKey, gateway)
+		if err != nil {
+			logger.Info("cannot get gateway", "error", err)
+			continue
+		}
+
+		routeList := &gatewayapiv1.HTTPRouteList{}
+		fields := client.MatchingFields{fieldindexers.HTTPRouteGatewayParentField: client.ObjectKeyFromObject(gateway).String()}
+		if err = m.opts.Client.List(ctx, routeList, fields); err != nil {
+			logger.Info("cannot list httproutes", "error", err)
+			continue
+		}
+		policies := policyKind.List(ctx, m.opts.Client, obj.GetNamespace())
+		if len(policies) == 0 {
+			logger.Info("no kuadrant policy possibly affected by the gateway related event")
+			continue
+		}
+		topology, err := kuadrantgatewayapi.NewTopology(
+			kuadrantgatewayapi.WithGateways([]*gatewayapiv1.Gateway{gateway}),
+			kuadrantgatewayapi.WithRoutes(utils.Map(routeList.Items, ptr.To[gatewayapiv1.HTTPRoute])),
+			kuadrantgatewayapi.WithPolicies(policies),
+			kuadrantgatewayapi.WithLogger(logger),
+		)
+		if err != nil {
+			logger.Info("unable to build topology for gateway", "error", err)
+			continue
+		}
+		index := kuadrantgatewayapi.NewTopologyIndexes(topology)
+		data := utils.Map(index.PoliciesFromGateway(gateway), func(p kuadrantgatewayapi.Policy) reconcile.Request {
+			policyKey := client.ObjectKeyFromObject(p)
+			logger.V(1).Info("kuadrant policy possibly affected by the gateway related event found")
+			return reconcile.Request{NamespacedName: policyKey}
+		})
+		requests = append(requests, data...)
 	}
 
-	if len(requests) == 0 {
-		logger.V(1).Info("no kuadrant policy possibly affected by the httproute related event")
+	if len(requests) != 0 {
+		return requests
 	}
 
+	policyKey, err := kuadrant.DirectReferencesFromObject(httpRoute, policyKind)
+	if err != nil {
+		return requests
+	}
```

Comment:
Should we log this `err` at some level in case something goes wrong for traceability?